Add metrics around resources with invalid configuration (#5710)
solo-io/gloo: docs/content/guides/observability/prometheus/metrics.md

More information on Gloo Edge's admin ports can be found [here]({{% versioned_link_path fromRoot="/introduction/observability/#grafana-and-prometheus" %}}).

> ### Monitoring the Status of Gloo Resources

Reviewer: ```suggestion
### Monitoring the configuration status of Gloo resources
```

When Gloo Edge fails to process a resource, we see that reflected in the resource's {{< protobuf name="core.solo.io.Status" display="Status">}}.

> Gloo Edge can also be configured to publish metrics which record the health status of resources.

Reviewer: ```suggestion
You can configure Gloo Edge to publish metrics that record the configuration status of resources.
```

changed health to configuration bc that's what we call it in the reference doc

> The status metrics can be enabled by specifying a resource type, along with the label(s) that should be applied to the metric, within the
> `observabilityOptions` in the Settings CRD. Consider the following, which will add metrics for virtual services and

Reviewer: ```suggestion
In the `observabilityOptions` of the Settings CRD, you can enable status metrics by specifying the resource type and any labels to apply
to the metric. Consider the following, which will add metrics for virtual services and
```

upstreams, both having labels that include the namespace and name of each individual resource:

```yaml

> observabilityOptions:

Author: We don't do any handling or even concern ourselves with the semantics of the labels selected, we simply grab the field out of the resource and stick it onto the metric.

If `{.metadata.namespace}` is used as a label value in the "two installations, separate namespaces" scenario, and both installations are writing a status metric for the same underlying resource, then both installations would use the same values for the labels. The installations could, however, have different opinions on the status of the resource. The label values relate to fields on the resource, and don't have anything to do with the installation. This code would simply look at the resource, grab whatever value is in `resource.metadata.namespace` and stick that onto the metric.

In this scenario, would each installation have their own prometheus server? IIUC, when setting metrics, they're just set in memory until something comes by, scrapes them, and persists them somewhere long-term. In order to avoid two installations writing metrics over top of one another, whoever is collecting the metrics is the person who is responsible for keeping the two separated.  This applies for all metrics, not just statuses.

  configStatusMetricLabels:
    Upstream.v1.gloo.solo.io:

>       labelToPath:

Author: The criteria for what we can set the label value on the metric to isn't all that restrictive: https://opencensus.io/tag/value/

I don't believe there's any restrictions on what's "valid" to pull out of the CR, as long as it meets the tag value criteria. Grabbing fields from the spec shouldn't be an issue. It really just comes down to what customers would find valuable to have. For the most part I would think namespace and name are enough to identify the resource experiencing issues

        name: '{.metadata.name}'
        namespace: '{.metadata.namespace}'
    VirtualService.v1.gateway.solo.io:
      labelToPath:
        name: '{.metadata.name}'
        namespace: '{.metadata.namespace}'
```

> After completing the [Hello World]({{% versioned_link_path fromRoot="/guides/traffic_management/hello_world/" %}}) 

Reviewer: ```suggestion
After completing the [Hello World guide]({{% versioned_link_path fromRoot="/guides/traffic_management/hello_world/" %}}) 
```

to generate some resources, the metrics defined above can be found at: <http://localhost:9091/metrics>. If the port
forwarding is directed towards the gloo pod, the `default-petstore-8080` upstream can be seen in a healthy state:
```
validation_gateway_solo_io_upstream_config_status{name="default-petstore-8080",namespace="gloo-system"} 0
```

If the port forwarding is then switched to the gateway pod, the metric defined for virtual services can be found

> on the same metrics endpoint (after a refresh):

Reviewer: a refresh of what?

Author: The page in a browser

Author: I went ahead and rephrased it, let me know if that works for you.

Thanks for the review!
